qcodes/dataset/data_set_tools.py

```diff
@@ -28,25 +29,32 @@ def merge(datasets: Sequence[Union[int, DataSet]]) -> DataSet:
             datasets.insert(datasets.index(ds), load_by_id(ds))
             datasets.remove(ds)
 
+    dsets = cast(List[DataSet], datasets)
+
     # Step 1: verify that all parameters match
-    ds_pairs = combinations(datasets, 2)
+    ds_pairs = combinations(dsets, 2)
     for pair in ds_pairs:
         if not _params_are_the_same(pair[0], pair[1]):
             raise ValueError(f'Can not merge datasets {pair[0]} and {pair[1]},'
                              ' parameters are not the same in those datasets.')
 
     # Step 2: construct a new dataset
-    new_ds_name = f'merge_of_{"_".join([str(ds.run_id) for ds in datasets])}'
+    new_ds_name = f'merge_of_{"_".join([str(ds.run_id) for ds in dsets])}'
     new_ds = new_data_set(name=new_ds_name,
-                          specs=datasets[0].get_parameters())
+                          specs=dsets[0].get_parameters())
 
     # Step 3: fill in the data of the old datasets
     results = []
-    for ds in datasets:
+    snapshots = {}
+    for ds in dsets:
         params = ds.get_parameters()
         data = ds.get_data(*params)
         for data_point in data:
             results.append({p.name: val for p, val in zip(params, data_point)})
+        snapshot = ds.get_metadata('snapshot')
+        snapshot = "{}" if snapshot is None else snapshot
+        snapshots[ds.run_id] = json.loads(snapshot)
     new_ds.add_results(results)
+    new_ds.add_metadata('parent_snapshots', json.dumps(snapshots))
```

Comment:
Beware that this way you are adding a new column to the database. `add_metadata` is implemented in such a way that a `tag` is actually a completely new column in the `runs` table. I am a bit concerned about this fact for the reason that the structure of columns should probably be restricted to 'smth' (because it is safer, isn't it; but yes, less flexible; but who needs THIS flexibility...). This is why my proposal was to have what you called `parent_snapshots` inside the `snapshot` part of metadata. What do you think?

Reply:
Yes, I am aware of that, and it has also been bugging me, but at the same time we have no idea what metadata we are eventually going to need, so for now I think dynamically adding columns is acceptable. 

I considering putting this in as a `snapshot`, but I am worried that snapshot-reading tools will break unless the snapshot has a particular format.